[WIP] Add protection tools

ManfredKarrer opened the pull request:
Merges https://github.com/bisq-network/bisq/pull/2765 and https://github.com/bisq-network/bisq/pull/2768

ManfredKarrer committed:
Merge branch 'fix-trade-peers-account-age' into delay-payout-if-fait-buyer-has-new-account

ManfredKarrer committed:
Fix wrong display of ages with 1 at the end (e.g. 31 days)

31 days was displayed as 31 day.

ManfredKarrer committed:
Add min required buyers account age

Implements https://github.com/bisq-network/proposals/issues/77

If fiat trade the buyer need to have an min required account age,
otherwise the payout will be delayed.

ManfredKarrer committed:
Add warning popup to offer book

ManfredKarrer committed:
Remove rounded fiat amounts popup

I think the features is now well known and can be removed.
The account age check will add a nde popup so we would get more popup
hell if we leave that one...

ManfredKarrer committed:
Add popup to offer maker

If offer maker for buying BTC with Fiat has an account selected which is
less then the required account age for that method he gets a
popup displayed.

ManfredKarrer committed:
Refactor

ManfredKarrer committed:
Improve text

ManfredKarrer committed:
Handle trade period (WIP)

ManfredKarrer committed:
Add more keys

ManfredKarrer committed:
Refactor: Move accountAgeWitness classes to new package

ManfredKarrer committed:
Refactor: Cleanup, no functionality change

ManfredKarrer committed:
Add SignedWitness domain (WIP)

ManfredKarrer committed:
Merge branch 'master' into delay-payout-if-fait-buyer-has-new-account

# Conflicts:
#	desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/PendingTradesDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/steps/seller/SellerStep3View.java

ManfredKarrer committed:
Update strings

ManfredKarrer committed:
Merge branch 'master' into sign-account-usage

# Conflicts:
#	core/src/main/java/bisq/core/account/witness/AccountAgeWitnessService.java
#	desktop/src/main/java/bisq/desktop/components/PeerInfoIcon.java

ManfredKarrer committed:
Fix wrong imports

ManfredKarrer committed:
Refactoring: Move classed to new package

Pure technical refactoring, no functionality change.

ManfredKarrer committed:
Add Account score domain (WIP)

Move methods for delayed payout to AccountScore.

ManfredKarrer committed:
Refactor: Move methods to domain (WIP)

ManfredKarrer committed:
Move account age related code to AccountCreationAgeService

We use AccountScoreService as main class which constructs the
parameters based on several protection tools. Currently there is only
account age based protection.

ManfredKarrer committed:
Fix delay function, refactorings

ManfredKarrer committed:
Rename method

ManfredKarrer committed:
Add methods for getting the AccountScoreCategory

ManfredKarrer committed:
Merge branch 'sign-account-usage' into add-protection-tools

# Conflicts:
#	core/src/main/java/bisq/core/app/BisqSetup.java
#	core/src/main/java/bisq/core/app/misc/AppSetupWithP2P.java
#	core/src/main/java/bisq/core/app/misc/AppSetupWithP2PAndDAO.java
#	core/src/main/java/bisq/core/offer/OfferUtil.java
#	core/src/main/java/bisq/core/proto/CoreProtoResolver.java
#	core/src/main/java/bisq/core/proto/persistable/CorePersistenceProtoResolver.java
#	core/src/main/java/bisq/core/trade/Trade.java
#	core/src/main/java/bisq/core/trade/TradeManager.java
#	core/src/main/java/bisq/core/trade/TradeModule.java
#	core/src/main/java/bisq/core/trade/protocol/ProcessModel.java
#	core/src/main/java/bisq/core/trade/protocol/tasks/VerifyPeersAccountAgeWitness.java
#	desktop/src/main/java/bisq/desktop/components/PeerInfoIcon.java
#	desktop/src/main/java/bisq/desktop/components/PeerInfoIconSmall.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/AdvancedCashForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/AliPayForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/AssetsForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/BankForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/CashDepositForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/ChaseQuickPayForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/ClearXchangeForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/F2FForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/FasterPaymentsForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/GeneralAccountNumberForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/GeneralBankForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/GeneralSepaForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/HalCashForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/InteracETransferForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/MoneyBeamForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/MoneyGramForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/NationalBankForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/PaymentMethodForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/PerfectMoneyForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/PopmoneyForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/PromptPayForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/RevolutForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/SameBankForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/SepaForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/SepaInstantForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/SpecificBankForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/SwishForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/USPostalMoneyOrderForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/UpholdForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/WeChatPayForm.java
#	desktop/src/main/java/bisq/desktop/components/paymentmethods/WesternUnionForm.java
#	desktop/src/main/java/bisq/desktop/main/MainViewModel.java
#	desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsDataModel.java
#	desktop/src/main/java/bisq/desktop/main/account/content/altcoinaccounts/AltCoinAccountsView.java
#	desktop/src/main/java/bisq/desktop/main/account/content/fiataccounts/FiatAccountsDataModel.java
#	desktop/src/main/java/bisq/desktop/main/account/content/fiataccounts/FiatAccountsView.java
#	desktop/src/main/java/bisq/desktop/main/market/offerbook/OfferBookChartViewModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/MutableOfferDataModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/createoffer/CreateOfferDataModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/offerbook/OfferBookViewModel.java
#	desktop/src/main/java/bisq/desktop/main/offer/takeoffer/TakeOfferDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/closedtrades/ClosedTradesViewModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/editoffer/EditOfferDataModel.java
#	desktop/src/main/java/bisq/desktop/main/portfolio/pendingtrades/PendingTradesViewModel.java
#	desktop/src/test/java/bisq/desktop/main/offer/createoffer/CreateOfferViewModelTest.java
#	desktop/src/test/java/bisq/desktop/main/portfolio/editoffer/EditOfferDataModelTest.java

ManfredKarrer committed:
Add arguments

ManfredKarrer committed:
Apply organize imports

Pure technical refactoring, no functional code change

ManfredKarrer committed:
Rename package

Pure technical refactoring, no functional code change

ManfredKarrer committed:
Handle AccountScoreCategory methods depending on signed witness state

ManfredKarrer committed:
Add verify check

ManfredKarrer committed:
Apply new delay function, add ScoreInfo class and methods

ManfredKarrer committed:
Add UI for score info

ManfredKarrer committed:
Fix deposit function

ManfredKarrer committed:
Adjust create offer popup with new scoreInfo data

ManfredKarrer committed:
Add popup for taker

ManfredKarrer committed:
Change buyer sec. deposit

ManfredKarrer committed:
Change PERM_DELAY to 15 days, adjust tests

ManfredKarrer committed:
Add requireAuthorizedTaker to offer and preferences

ManfredKarrer committed:
Add check and popup for offers requiring authorized takers

- Revert earlier changes which applied that check in the general
account check.

ManfredKarrer committed:
Adjust column width

ManfredKarrer committed:
Show restrictions based on payment method

ripcurlx committed:
Improve restrictions settings in create offer

ripcurlx commented on core/src/main/java/bisq/core/account/score/ScoreInfo.java:
```diff
@@ -0,0 +1,51 @@
+    private AccountScoreCategory accountScoreCategory;
+    private long accountAge;
+    private Optional<Long> signedTradeAge;
+    private Coin minDepositAsCoin; //TODO remove?
```
I think this will be gone as we are not able to force this from the maker's side - right?

ManfredKarrer replied:
Probably

ripcurlx commented on core/src/main/java/bisq/core/account/sign/SignedWitness.java:
```diff
@@ -0,0 +1,156 @@
+    private final byte[] signerPubKey;
+    private final byte[] witnessOwnerPubKey;
+    private final long date;
+    //TODO should we add trade amount?
```
Is the idea to add the trade amount to add this as a weight later on as well? Of course a 1000 USD trade might not be undetected compared to a 10 USD trade.

ManfredKarrer replied:
Not sure yet if needed. Simplicity might be better here.... but leave the comment open until 100% clear.

ripcurlx commented on core/src/main/java/bisq/core/account/sign/SignedWitness.java:
```diff
@@ -0,0 +1,156 @@
+    @Override
+    public boolean isDateInTolerance() {
+        // We don't allow older or newer then 1 day.
+        // Preventing forward dating is also important to protect against a sophisticated attack
```
The date is not part of the hash. Is this by intent?

ManfredKarrer replied:
Will think about it

ManfredKarrer replied:
Just added a comment:
  // Date is not added to hash to avoid that the same account with same sigs can be stored multiple 
        // times if date would differ

ripcurlx commented on core/src/main/java/bisq/core/account/sign/SignedWitnessService.java:
```diff
@@ -0,0 +1,257 @@
+            return arbitratorsSignedWitnessSet.stream().anyMatch(this::verify);
+        } else {
+            Set<SignedWitness> trustedPeerSignedWitnessSet = getTrustedPeerSignedWitnessSet(accountAgeWitness);
+            // We have some SignedWitness singed by any trusted peer and need to see if it is valid and has a
```
typo `singed` should be `signed`

ManfredKarrer committed:
Merge pull request #8 from ripcurlx/create-offer-improvements

Improve restrictions settings in create offer

ManfredKarrer committed:
Add comment

ManfredKarrer committed:
Add trade amount to SignedWitness

Even if we would not use it initially it might be good to have it for
potential future use.

ManfredKarrer committed:
Rename put method to putIfAbsent to reflect behaviour better

ManfredKarrer committed:
Rename put method to putIfAbsent to reflect behaviour better

ManfredKarrer committed:
Add more comments

ripcurlx committed:
Add missing parameter

ripcurlx committed:
Change icons

ManfredKarrer committed:
Merge pull request #9 from ripcurlx/change-icons

Change icons

ManfredKarrer committed:
Extend trade protocol to support payout delay

ManfredKarrer committed:
Add missing class

ManfredKarrer committed:
Fix fiatReceivedDate handling

ManfredKarrer committed:
Rename PeerInfoWithTagEditor to PeerInfoPopup

ManfredKarrer committed:
Handle delayed payout in trade protocol (WIP)

ManfredKarrer committed:
Add verify task, improve UI

ManfredKarrer commented:
I am considering that implementation approach to be on hold. See https://github.com/bisq-network/proposals/issues/77#issuecomment-489200973 for my concerns with the approach of using a delayed payout.

ManfredKarrer commented:
That approach has been dropped in favor of other ideas (see https://github.com/bisq-network/proposals/issues/78, https://github.com/bisq-network/proposals/issues/83)

ManfredKarrer committed:
Handle payout delay in trade process (WIP)
